### PR TITLE
Add TrainingTrace

### PR DESCRIPTION
**Pull request by blazekadam**
- add `TrainingTrace` class
- use `TrainingTrace` to store training begin, end, epochs_done and exit_status
- expose some properties in `MainLoop`
- add `TrainingTrace` tests

**IMPORTANT: review and merge after merging #163**

### CODE REVIEW

File: cxflow/utils/training_trace.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    def save(self) -> None:
+        """
+        Save the training trace to :py:attr:`CXF_TRACE_FILE` file under the specified directory.
+
+        :raise ValueError: if no output directory was specified
+        """
+        if self._output_dir is None:
+            raise ValueError('Can not save TrainingTrace without output dir.')
+        config_to_file(self._trace, self._output_dir, CXF_TRACE_FILE)
```
Comment (petrbel) [inline cxflow/utils/training_trace.py]: `config_to_file` should be refactored. It handles arbitrary dict, not just config.

File: cxflow/utils/config.py [hunk 2]
```diff
@@ -26,7 +26,7 @@ def parse_arg(arg: str) -> typing.Tuple[str, typing.Any]:
     return key, value
 
 
-def load_config(config_file: str, additional_args: typing.Iterable[str]) -> dict:
+def load_config(config_file: str, additional_args: typing.Iterable[str] = ()) -> dict:
```
Comment (petrbel) [inline cxflow/utils/config.py]: spaces should be removed

File: cxflow/utils/training_trace.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    Lightweight wrapper object allowing to modify, save and load training trace variables
+    defined in :py:class:`TrainingTraceKeys`.
+    """
+
+    _ENUM_VALUES = {}
+    """
+    Mapping of :py:class:`TrainingTraceKeys` to ``Enum``s used for (de)serialization and checks of ``Enum`` values.
+    """
+
+    def __init__(self, output_dir: Optional[str] = None, autosave: bool = True):
```
Comment (petrbel) [inline cxflow/utils/training_trace.py]: spaces should be removed

File: cxflow/utils/training_trace.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        Create new ``TrainingTrace``.
+
+        :param output_dir: training output directory to save the trace to
+        :param autosave: auto save the trace after setting an item
+        """
+        self._output_dir = output_dir
+        self._autosave = autosave
+        self._trace = OrderedDict([(key.value, None) for key in TrainingTraceKeys])
+
+    def __setitem__(self, key: TrainingTraceKeys, item: object) -> None:
```
Comment (petrbel) [inline cxflow/utils/training_trace.py]: I'd prefer `typing.Any` instead of `object`

File: cxflow/utils/training_trace.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        assert isinstance(key, TrainingTraceKeys), 'TrainingTrace key must be instance of TrainingTraceKeys enum'
+        if key in TrainingTrace._ENUM_VALUES:
+            assert isinstance(item, TrainingTrace._ENUM_VALUES[key]), \
+                '`{}` item must be instance of `{}`'.format(key, TrainingTrace._ENUM_VALUES[key])
+            item = item.value  # map enum item to its value
+        self._trace[key.value] = item
+        if self._autosave:
+            self.save()
+
+    def __getitem__(self, key) -> Optional[object]:
```
Comment (petrbel) [inline cxflow/utils/training_trace.py]: I'd prefer `typing.Any` instead of `object`

Comment (petrbel) [reply cxflow/utils/training_trace.py]: btw missing `key` type-hint

File: cxflow/utils/training_trace.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    @staticmethod
+    def from_file(filepath: str):
+        """
+        Load training trace from the given ``filepath``.
+
+        :param filepath: training trace file path
+        :return: training trace
+        """
+        trace = TrainingTrace()
+        trace._trace = load_config(filepath)
```
Comment (petrbel) [inline cxflow/utils/training_trace.py]: we don't use this function only for loading the config, it should be renamed

File: cxflow/utils/training_trace.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        self._trace[key.value] = item
+        if self._autosave:
+            self.save()
+
+    def __getitem__(self, key) -> Optional[object]:
+        """
+        Get an item for the given key.
+        Only :py:class:`TrainingTraceKeys` are allowed as keys.
+        """
+        assert isinstance(key, TrainingTraceKeys), 'TrainingTrace key must be instance of TrainingTraceKeys enum'
```
Comment (petrbel) [inline cxflow/utils/training_trace.py]: `TrainingTrace` -> `` `TrainingTrace` ``

File: cxflow/utils/training_trace.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    """Program exit status."""
+
+
+class TrainingTrace:
+    """
+    Lightweight wrapper object allowing to modify, save and load training trace variables
+    defined in :py:class:`TrainingTraceKeys`.
+    """
+
+    _ENUM_VALUES = {}
```
[the remark this answers is not in the document]
Comment (blazekadam) [reply cxflow/utils/training_trace.py]: It is empty on purpose and yes it should contain mapping to types (Enums in particular) as the docstring suggests. It allows type safe extension of `TrainingTrace` with additional entries with Enum values.

**Pull request merged**